### PR TITLE
Improve custom entity handling and add more unit handling

### PR DESCRIPTION
**Pull request by Pho3niX90**
This commit improves the handling of custom entities, particularly focusing on ensuring the state is not null, undefined, or unknown. It also adds various system features such as conversion from a state to power and auto conversion from a state to kilowatts. In addition, units of electrical current, energy, and power were added to provide spatial and time recognizability context to the measures. This enhancement results in a more detailed and accurate representation of data. It also includes adjustments in energy conversion for better precision and clarity.

possibly fixes: #299

### CODE REVIEW

**Comment by slipx06** [comment]
Nice improvements. Working through it. Thanks

**Comment by slipx06** [comment]
Note to check logic when autoscale is false as it shows units in kWh
image
When true it shows W
image

It should default to the lowest units i.e. W?

File: src/inverters/dto/custom-entity.ts [hunk 1]
```diff
@@ -35,19 +35,34 @@ export interface CustomEntity extends HassEntity {
...
+        isValid: () => entity?.state !== null && entity.state !== undefined && entity.state !== 'unknown' || false,
         notEmpty: () => entity?.state !== '' || false,
         isNaN: () => Number.isNaN(entity?.state) || true,
         toPower: (invert?: boolean) => (entity.attributes?.unit_of_measurement || '').toLowerCase() === 'kw'
             ? Utils.toNum(((entity?.state || '0') * 1000), 0, invert)
             : Utils.toNum((entity?.state || '0'), 0, invert) || 0,
-        toString: () => entity?.state?.toString() || ''
+        toPowerString: (scale?: boolean, decimals?: number, invert?: boolean) =>
+            scale ?
+                Utils.convertValueNew(entity?.state, entity?.attributes?.unit_of_measurement, decimals || 1) :
```
Comment (slipx06) [inline src/inverters/dto/custom-entity.ts:63]: This does not appear to be working properly. When scale is true it does not auto convert the values. Tested on pv1 and pv2 power

File: src/helpers/utils.ts [hunk 2]
```diff
@@ -22,85 +24,32 @@ export class Utils {
...
+            if (Math.abs(numberValue) >= rule.threshold) {
+                const convertedValue = (numberValue / rule.divisor).toFixed(rule.decimal || decimal);
+                return `${convertedValue} ${rule.targetUnit}`;
             }
-        } else {
-            // If the unit is not in the allowed units, return the numeric value without unit
-            return Math.round(numberValue);
         }
+
+        return `${numberValue.toFixed(decimal)} ${unit}`;
```
Comment (slipx06) [inline src/helpers/utils.ts:42]: The original function allowed the return of an empty unit if not provided.

File: src/const.ts [hunk 3]
```diff
@@ -28,3 +28,52 @@ export const enum SensorDeviceClass {
...
+}
+
+type ConversionRule = {
+    threshold: number;
+    divisor: number;
+    targetUnit: UnitOfEnergy;
+    decimal?: number;
+};
+
+export const unitOfEnergyConversionRules: Record<UnitOfEnergy, ConversionRule[]> = {
```
Comment (slipx06) [inline src/const.ts:63]: The original function also converted W to kW to MW. This is probably the reason it is not aut scaling all the power values

File: src/inverters/dto/custom-entity.ts (hunk 1 above)
Comment (slipx06) [reply src/inverters/dto/custom-entity.ts:63]: I've updated the function to include power values

File: src/const.ts (hunk 3 above)
Comment (slipx06) [reply src/const.ts]: Updated to include power conversion

**Comment by slipx06** [comment]
Have applied changes and checked both full and compact cards. Please can you also confirm everything working on your end.

**Comment by Pho3niX90** [comment]
Soz just seeing everything now, will quickly test on my end

**Comment by Pho3niX90** [comment]
Okay all seems good my side as well

**Pull request merged**